### PR TITLE
design document: threat model

### PR DESCRIPTION
**Pull request by tgeoghegan**
Initial threat model, enumerating the system's actors and their
capabilities, as well as current or hypothetical mitigations.

### CODE REVIEW

**Review by chris-wood — CHANGES_REQUESTED** [review]
This is a great start! I think my only blocking comment is around the network adversary.

File: design-document.md [hunk 1]
```diff
@@ -106,6 +123,132 @@ to achieve these goals:
...
+
+1. Unshared data. Clients are the only actor that can ever see the original
+   data.
+1. Unencrypted data shares.
+1. Client identity (optionally).
+
+##### Capabilities
+
+1. Individual clients can trivially and undetectably defeat the privacy property
+   by leaking data outside of the Prio system. Accordingly, such attacks are
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:146]: I'm making a distinction between the user and the client here (I think my use of "individual" is confusing). I'm imagining a scenario where an application vendor might adopt Prio to elicit trust from users, but ship client software that covertly leaks user data over some non-Prio channel. 

I'm not sure we can do anything about it, unless the model is extended to include a trusted OS that a client runs in, which could attest to various properties of the client (e.g., measurements of reproducible builds, or attest to a log of the client's network activity so that external observers can verify that there are no unexpected connections to `secret-cleartext-metrics.appvendor.com`).

I'm going to reword this to explicitly separate the case of an individual user/client disclosing their own data (which as you say does not compromise anyone else's privacy) and the case of a client used by many users leaking data (which is out of scope in the current model).

File: design-document.md [hunk 2]
```diff
@@ -106,6 +123,132 @@ to achieve these goals:
...
+   to emit aggregation parts.
+
+##### Mitigations
+
+1. Infrastructure diversity. Prio deployments should ensure that aggregators'
+   security do not have a common point of failure. For instance, if all
+   participating aggregators stored unencrypted data shares on a single cloud
+   object storage service, then that cloud vendor would be able to decrypt and
+   reassemble all the data shares and defeat privacy.
+1. Implementation diversity. Similarly to infrastructure diversity, diversity in
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:196]: I agree that this one isn't specific to Prio but rather would apply to any distributed computing system, so I'll strike it.

File: design-document.md [hunk 3]
```diff
@@ -106,6 +123,132 @@ to achieve these goals:
...
+1. Individual aggregators may compromise availability of the system by refusing
+   to emit aggregation parts.
+
+##### Mitigations
+
+1. Infrastructure diversity. Prio deployments should ensure that aggregators'
+   security do not have a common point of failure. For instance, if all
+   participating aggregators stored unencrypted data shares on a single cloud
+   object storage service, then that cloud vendor would be able to decrypt and
+   reassemble all the data shares and defeat privacy.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:195]: The point of Prio, as I see it, is that you need to subvert all of the aggregators in order to defeat privacy. But if all the aggregators are using the same cloud storage vendor, then the attacker no longer needs to subvert _n_ different aggregators -- they just need to subvert a single cloud vendor. So users are back to hoping that a single entity will not have any breaches, will not get any national security letters and will consistently comply with its own privacy policy, which I think undermines Prio.

Reading this back, I don't like the sentence "Prio deployments should ensure that aggregators' security do not have a common point of failure.". I'll try to find a better way to express this.

File: design-document.md [hunk 4]
```diff
@@ -106,6 +123,132 @@ to achieve these goals:
...
+   minimize the impact of false data.
+1. Many deployments could allow anyone to contribute data shares to aggregators,
+   but others could require that clients establish a trusted identity before
+   they may contribute data. A variety of techniques providing different degrees
+   of assurance could be employed to establish trust in clients, ranging from
+   whitebox cryptography to reproducible builds and remote attestation of
+   software measurements from a trusted computing base.
+
+[[OPEN ISSUE: The problem of clients sending bogus data exists in any metrics
+system, orthogonally to Prio, so perhaps it is out of scope.]]
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:168]: I'm not so sure, because at least one mitigation could be applied within the Prio protocol: requiring client identities. If clients had to authenticate to aggregators (or to a trusted authenticator who would then relay anonymized data shares to aggregators), then the system could detect clients sending excessive contributions and reject them.

File: design-document.md [hunk 5]
```diff
@@ -106,6 +123,132 @@ to achieve these goals:
...
+1. Data share decryption keys.
+1. Aggregation parts.
+1. Aggregator identity.
+
+##### Capabilities
+
+1. Aggregators may trivially and undetectably defeat the robustness of the
+   system by emitting bogus aggregation shares. Accordingly, such attacks are
+   outside of the threat model.
+1. If client identities are used, aggregators may weaken the anonymity of the
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:184]: I would phrase it as "if a client reveals identifying information to aggregators, such as via client auth, then aggregators have this capability".

File: design-document.md [hunk 6]
```diff
@@ -106,6 +123,132 @@ to achieve these goals:
    function output itself. {{questions-and-params}} discusses the leakage profiles of 
    various aggregation functions in more detail.
 
+### Threat model
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:126]: I think you're right that there should be a section for an attacker on the network. However I find it hard to reason about a threat like the one in your example because we haven't yet sketched out how aggregator discovery works. But certainly we can already call out the necessity of confidential and mutually authenticated links between participating servers.

File: design-document.md [hunk 7]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+       would be able to reassemble all the data shares and defeat privacy.
+1. Running the protocol over multiple subsets of the available aggregators
+   chosen so that no aggregator appears in all subsets and accepting any result
+   would prevent an individual aggregator from compromising availability.
+1. Running the protocol over multiple subsets of the available aggregators
+   chosen so that no aggregator appears in all subsets and requiring that the
+   results agree would prevent an individual aggregator from compromising
+   robustness
+     * This could also allow identifying of defective aggregators, by finding
+       the aggregator that appears in every subset whose results were wrong.
```
Comment (cjpatton) [inline design-document.md:210]: ^ Does this defeat the bogus aggregation share attack? I'm not sure anything does. There may be other, more subtle attacks it detects, but given how easy it is for an aggregator to submit a bogus aggregation share, I don't think it matters much. (See comment in "Future work and possible extensions".)

However, detecting bad aggregators is a really good idea. We'll need to flesh this out later.

File: design-document.md [hunk 8]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+
+#### Leader
+
+The leader is also an aggregator, and so all the assets, capabilities and
+mitigations available to aggregators also apply to the leader.
+
+##### Capabilities
+
+1. Data validity proof verification. The leader can forge proofs and collude
+   with a malicious client to trick aggregators into aggregating invalid data.
```
Comment (cjpatton) [inline design-document.md:220]: This capability is only marginally stronger than the bogus aggregation share attack, IMO.

File: design-document.md [hunk 9]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+     * This could also allow identifying of defective aggregators, by finding
+       the aggregator that appears in every subset whose results were wrong.
+
+#### Leader
+
+The leader is also an aggregator, and so all the assets, capabilities and
+mitigations available to aggregators also apply to the leader.
+
+##### Capabilities
+
```
Comment (cjpatton) [inline design-document.md:218]: Another capability to call out: The leader can try to shrink the anonymity set by requesting aggregation shares early (cf. https://github.com/abetterinternet/prio-documents/issues/15).

File: design-document.md [hunk 10]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+1. Data validity proof verification. The leader can forge proofs and collude
+   with a malicious client to trick aggregators into aggregating invalid data.
+1. Relaying messages between aggregators. The leader can compromise availability
+   by dropping messages.
+
+##### Mitigations
+
+1. Rather than having all aggregators trust the leader to validate proofs, each
+   aggregator could exchange proof shares with every other aggregator, enabling
+   each aggregator to independently verify the proofs.
```
Comment (cjpatton) [inline design-document.md:228]: I think what's meant by this is that the aggregators could exchange verification shares so that each aggregator can run the decision algorithm locally. Is that right?

I agree that this would help mitigate the leader's additional capabilities. However, doing so would require a secure channel between each pair of aggregators. I think this cost is too high.

File: design-document.md [hunk 11]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+1. Running the protocol over multiple subsets of the available aggregators and
+   choosing different leaders for each run would prevent any individual leader
+   from compromising availability.
+
+#### Collector
+
+##### Capabilities
+
+1. Advertising shared configuration parameters (e.g., minimum thresholds for
+   aggregations, polynomial identity test values, arithmetic circuits).
```
Comment (cjpatton) [inline design-document.md:238]: In [BBC+19] (and hence libprio-rs) the concept of "polynomial identity test" is generalized to distributed evaluation of a PCP over joint randomness. Hence, I would use the term "joint randomness" instead of "polynomial identity test values".

To your point: We haven't yet decided how the joint randomness is chosen. It could very well be the collector, but this is TBD. (See "Consensus protocol" below.) Also, I would add "designation the leader role" to this list.
But

File: design-document.md [hunk 12]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+1. Collectors may trivially defeat availability by discarding aggregation
+   parts submitted by aggregators.
```
Comment (cjpatton) [inline design-document.md:240]:
```suggestion
   shares submitted by aggregators.
```

File: design-document.md [hunk 13]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+
+##### Mitigations
+
+1. Aggregators should refuse shared parameters that are trivially insecure
+   (i.e., aggregation threshold of 1 contribution).
```
Comment (cjpatton) [inline design-document.md:245]: The parameters need to be authenticated and enforced by each aggregator, but I think it's up to the collector to choose how large the anonymity set should be.

File: design-document.md [hunk 14]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+        information about user activity.
+          * For example, if a user opts into a new feature, and the client
+            immediately reports this to aggregators, then just by observing
+            network traffic, the attacker can infer what the user did.
+     1. Observation of message size could allow the attacker to learn how much
+        data is being submitted by a client.
+          * For example, if the attacker observes an encrypted message of some
+            size, they can infer the size of the plaintext, plus or minus the
+            cipher block size. From this they may be able to infer which
+            aggregations the user has opted into or out of.
```
Comment (cjpatton) [inline design-document.md:271]: The size of the input is a function of the data type, not the data itself. In particular, all clients will submit the same amount of data for a given aggregation job.

File: design-document.md [hunk 15]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+
+1. All messages exchanged between participants in the system should be
+   encrypted.
+1. All messages exchanged between aggregators, the collector and the leader
+   should be mutually authenticated so that network attackers cannot impersonate
+   participants.
+1. Clients should be required to submit data at regular intervals so that the
+   timing of individual messages does not reveal anything.
+1. A fixed-length encoding should be used for data shares. Additionally, clients
+   should submit dummy data even for aggregations the user has not opted into.
```
Comment (cjpatton) [inline design-document.md:285]: Aggregation isn't well-defined if input lengths vary, so I think the first part of this mitigation is irrelevant.

File: design-document.md [hunk 16]
```diff
@@ -115,6 +293,50 @@ collude and break soundness. Is this a contingency we need to address? There are
...
+
+In this section we discuss attacks that are not considered in the above threat
+model, and suggest mitigations that could be incorporated into implementations
+of this protocol or future revisions of this specfication.
+
+#### Client authentication
+
+Attackers can impersonate Prio clients and submit large amounts of valid but
+false data in order to spoil aggregations. Deployments could require that
+clients present a trusted identity before they may contribute data. For example,
```
Comment (cjpatton) [inline design-document.md:306]:
```suggestion
false data in order to spoil aggregations. Deployments could require client
authentication. For example,
```

File: design-document.md [hunk 17]
```diff
@@ -115,6 +293,50 @@ collude and break soundness. Is this a contingency we need to address? There are
...
+by requiring data submissions to be signed with a key trusted by aggregators. A
+variety of techniques providing different degrees of assurance could be employed
+to establish trust in clients, ranging from whitebox cryptography to public key
+infrastructures. However some deployments may opt to accept the risk of false
+data to avoid having to figure out how to distribute trusted identities to
+clients.
```
Comment (cjpatton) [inline design-document.md:312]: Another authentication mechanism: The leader might just require the user to login beforehand.

I'm curious how whitebox crypto might be used here?

File: design-document.md [hunk 18]
```diff
@@ -115,6 +293,50 @@ collude and break soundness. Is this a contingency we need to address? There are
...
+
+#### Client attestation
+
+In the current threat model, servers participating in the protocol have no
+insight into the activities of clients except that they have uploaded data into
+a Prio aggregation, meaning that clients could covertly leak a user's data into
+some other channel which compromises privacy and anonymity. If we introduce the
```
Comment (cjpatton) [inline design-document.md:319]:
```suggestion
some other channel which compromises privacy. If we introduce the
```

As I mentioned above, privacy implies anonymity as the terms are defined above. (You may disagree?)

File: design-document.md [hunk 19]
```diff
@@ -115,6 +293,50 @@ collude and break soundness. Is this a contingency we need to address? There are
 techniques in [BBC+19] that account for this; we need to figure out if they're
 practical.]]
 
+### Future work and possible extensions
```
Comment (cjpatton) [inline design-document.md:296]: An attacker capability explicitly considered in [BBC+19], but not in [GB17], is that a coalition of verifiers may attempt to convince the honest verifiers that an invalid data is valid. This doesn't matter much in our setting, since the bogus aggregation share attack descrbed above is equally effective and much easier to pull off. It would matter, however, if the aggregator were divided into two entities: one which validates input shares (the verifier); and another that accumulates them and emits the aggregation share. I think it's worth noting this somewhere, probably here.

**Review by cjpatton — COMMENTED** [review]
This is super useful, thank you for getting this started. My only high-level comment is that this analysis doesn't completely settle the question of whether we care if a malicious server colludes with a malicious client in order to corrupt the final output. As I mention below, this threat model is only briefly considered in [BBC+19] (and not at all in the original Prio paper, [GB17]); and because a malicious aggregator could always submit a bogus aggregate share, this model doesn't seem (to me anyway) relevant for our deployment.

I think this PR should explicitly remove this threat model from scope. We should assume that aggregators do their very best to validate and accumulate data shares correctly. That said, we should make it a goal to detect misbehaving aggregators. However, the target there needs to be detecting bogus shares.

File: design-document.md [hunk 20]
```diff
@@ -13,6 +13,22 @@
    aggregators.
 1. Leader: A distinguished aggregator that coordinates input validation and data
    aggregation.
+1. Data: The original data emitted by a client, before any encryption or secret
+   sharing scheme is applied.
+1. Data share: one of the _n_ shares output by feeding a datum into a linear
+   secret sharing scheme. Each share is to be transmitted to one of the _n_
+   participating aggregators.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:20]: That's a good idea. I keep finding myself writing "original data" or "unshared data" to disambiguate, so I'll adopt your "input" instead.

File: design-document.md [hunk 21]
```diff
@@ -13,6 +13,22 @@
...
+1. Invalid data: Data which cannot be represented in the affine-aggregatable
+   encoding chosen by participants. e.g., a UTF-8 string if the chosen encoding
+   is a vector of 16 bits.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:23]: My phrasing here was based on a misreading of the 2017 paper. I'll rewrite this to make it plain that invalid data is simply data for which the validity predicate does not hold.

File: design-document.md [hunk 22]
```diff
@@ -95,6 +111,7 @@ an aggergation function F:
    and nothing else. 
 1. Robustness. The adversary can influence the output of F only by reporting false 
    (untruthful) data. The output cannot be influenced in any other way.
+1. Anonymity. The adversary cannot learn which client submitted which data value.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:114]: The [2017 paper](https://crypto.stanford.edu/prio/paper.pdf) calls out privacy and anonymity as distinct security properties (section 2). Anonymity is the adversary not knowing which client submitted which input, privacy is the adversary not knowing what the inputs are. The reason I added anonymity to this list is because I think that in the setting where aggregators talk directly to clients, privacy is preserved but client anonymity is weakened since the aggregator can identify clients e.g. by IP. Further we can sketch out a mitigation -- an anonymizing proxy -- so I think it's worth discussing the anonymity property.

File: design-document.md (hunk 8 above)
Comment (tgeoghegan) [reply design-document.md:220]: Plus, HCG clarified for us yesterday that any aggregator, leader or not, can collude with malicious clients to achieve the same thing. I'll rewrite this.

File: design-document.md [hunk 23]
```diff
@@ -13,6 +13,22 @@
...
+   participating aggregators.
+1. Invalid data: Data which cannot be represented in the affine-aggregatable
+   encoding chosen by participants. e.g., a UTF-8 string if the chosen encoding
+   is a vector of 16 bits.
+1. False data: Data which is valid under the chosen affine-aggregatable encoding
+   but is not truthful. e.g., if the data being gathered is whether or not users
+   have clicked on a particular button, a client could report clicks when none
+   occurred.
+1. Aggregation: a statistical aggregation over a body of data which is of
+   interest to a collector.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:29]: Good point! I chose "output" for symmetry with your other good suggestion of "input".

File: design-document.md [hunk 24]
```diff
@@ -106,6 +123,167 @@ to achieve these goals:
...
+   data shares.
+1. Infrastructure diversity. Prio deployments should ensure that aggregators do
+   not have common dependencies that would enable a single vendor to reassemble
+   data.
+     * For example, if all participating aggregators stored unencrypted data
+       shares on the same cloud object storage service, then that cloud vendor
+       would be able to reassemble all the data shares and defeat privacy.
+1. Running the protocol over multiple subsets of the available aggregators
+   chosen so that no aggregator appears in all subsets and accepting any result
+   would prevent an individual aggregator from compromising availability.
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply design-document.md:204]: I think this isn't something we'd tackle in an initial deployment of Prio anyway, so I wrote up #22 on this topic, and will move this mitigation into future work and possible extensions.

File: design-document.md (hunk 17 above)
Comment (tgeoghegan) [reply design-document.md:312]: The notion with whitebox is that you could use it to distribute a shared secret to all "legitimate" instances of your client, and so having a message HMACed with a whiteboxed key gives you some kind of assurance on the provenance of the message. But thinking about it further, I've never been a big fan of whitebox, and mentioning it here isn't particularly helpful to explaining Prio, so I'll just strike it.

File: design-document.md (hunk 19 above)
Comment (tgeoghegan) [reply design-document.md:296]: I think the leader in our design is the verifier you describe, no? What HCG argued to us on 3/24 was that a verifier colluding with a malicious client doesn't have any capability that isn't already available to any aggregator colluding with a malicious client. I think what you're sketching out is a further variation where the leader is _only_ a verifier and not an aggregator?

File: design-document.md (hunk 14 above)
Comment (tgeoghegan) [reply design-document.md:271]: For any individual aggregation, yes, but imagine if a client supports dozens or hundreds of individual aggregations, and users can opt out of any of them. If the client only transmits input shares for the aggregations the client has opted into, then the volume of traffic observable by the adversary would change. Hence I argue that a client should always transmit the same amount of data, regardless of which aggregations a user has consented to.

File: design-document.md (hunk 13 above)
Comment (tgeoghegan) [reply design-document.md:245]: I agree, but the intent here is that aggregators should implement basic checks independent of the configuration of any aggregation to prevent abuse.

File: design-document.md (hunk 10 above)
Comment (tgeoghegan) [reply design-document.md:228]: HCG convinced us (or at least me!) that it's OK to have a designated proof verifier because the verifier is no more powerful than any aggregator that colludes with a malicious client, so I'm going to remove this item.

File: design-document.md (hunk 14 above)
Comment (cjpatton) [reply design-document.md:264]: Works for me.

File: design-document.md (hunk 19 above)
Comment (cjpatton) [reply design-document.md:289]: > I think the leader in our design is the verifier you describe, no? 

No. Here I'm imagining if any aggregator were split into two services: one which validates input shares and another that accumulates them.

In any case, the point here is moot I think. We will assume that the aggregators execute the input-validation correctly (this is setting I of [BBC+19].)

**Review by cjpatton — APPROVED** [review]
LGTM modulo editorial things.

**Review by chris-wood — APPROVED** [review]
